src/MonoMod.ILHelpers/MonoMod.ILHelpers.il

```diff
@@ -34,24 +34,38 @@
     
     .method public hidebysig static !!T& ObjectAsRef<T>(object obj) cil managed aggressiveinlining
     {
+#ifdef netcoreapp
+        ldarg obj
+        conv.u
+        ret
+#else
         .locals (
-            object pinned pin
-#ifndef netcoreapp
-            , !!T& ref
+            object pinned pin,
+            !!T& finalRef
+#ifdef NET35
+            , !!T** refPtr
 #endif

```

Comment:
I'm not a huge fan of gating this behavior behind NET35. I'm not actually sure when the issue is fixed; it's possible its not fixed until sometime after Mono transitioned to .NET Framework 4.x; I'd like to at least double check that, or otherwise always do the extra casting.